### PR TITLE
Embed: IPC (ready signaling, navigation)

### PR DESCRIPTION
**Pull request by nl0**
# Description

- support customizing display and behavior for s3 object link button
- send "ready" message
- send and receive "navigate" messages
- send "error" messages
- support setting initial route
- use nonce and origin
- add docs for embed
- adjust debug harness
  - add message log
  - add route field
  - add "navigate to" button

# TODO

- [x] Documentation
    - [x] JavaScript: basic explanation and screenshot of new features -- WIP
- [x] [Changelog](CHANGELOG.md) entry

### CODE REVIEW

File: docs/Catalog/Embed.md [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+      if (!iframeRef.current) return
+      // origin must be sent as a second parameter to enable cross-origin message passing
+      iframeRef.current.contentWindow.postMessage(msg, EMBED_ORIGIN)
+    },
+    [iframeRef],
+  )
+
+  const initialize = React.useCallback(() => {
+    postMessage({
+      type: 'init',
```
Comment (fiskus) [inline docs/Catalog/Embed.md:41]: Maybe highlight in comments which parameters are required and which are optional

File: catalog/app/embed/ipc.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+export function useMessageHandler(fn: (data: object) => void) {
+  const handleMessage = React.useCallback(
+    (e: MessageEvent) => {
+      if (
+        e.source !== window.parent ||
+        (PARENT_ORIGIN !== '*' && e.origin !== PARENT_ORIGIN) ||
+        !e.data?.type
+      )
+        return
+      fn(e.data)
```
Comment (fiskus) [inline catalog/app/embed/ipc.ts:31]: I think, it's worth using `{}` brackets here to help read

File: catalog/app/embed/Embed.js [hunk 3]
```diff
@@ -136,36 +138,40 @@ function BucketLayout({ bucket, children }) {
 }
 
 function useInit() {
+  const messageParent = ipc.useMessageParent()
```
Comment (fiskus) [inline catalog/app/embed/Embed.js:141]: `messageToParent` or `messageFromParent` could be more convinient

**Review by fiskus — APPROVED** [review]

File: docs/Catalog/Embed.md (hunk 1 above)
Comment (nl0) [reply docs/Catalog/Embed.md:41]: i've added this info to the API reference section

File: catalog/app/embed/Embed.js (hunk 3 above)
Comment (nl0) [reply catalog/app/embed/Embed.js:141]: "message" here is a verb, i.e. "to message one's parent". maybe it makes sense to use some other wording here, e.g. "notify" or smth, to distinguish sent and received messages ("command" vs "notification")

**Comment by nl0** [comment]
documentation is ready for review / adjustments -- @akarve pls have a look

File: docs/Catalog/Embed.md [hunk 4]
```diff
@@ -0,0 +1,349 @@
+The Quilt catalog has a special (limited) S3 browser UI that can be embedded into any web app using an iframe -- "Embed" further on.
```
Comment (akarve) [inline docs/Catalog/Embed.md:1]:
```suggestion
The Quilt catalog has limited-feature S3 browser called Embed, an embeddable iframe.
```

File: docs/Catalog/Embed.md [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+This embeddable browser is served off the main catalog server under `/__embed` route.
+
+![](../imgs/embed-example.png)
+
+
+## Usage
+
+Here's an example code written in plain JavaScript + React outlining the basic use cases:
```
Comment (akarve) [inline docs/Catalog/Embed.md:9]:
```suggestion
Here's a sample code snippet in plain JavaScript + React that outlines the basic use cases:
```

File: docs/Catalog/Embed.md [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+      //   url: string -- url / route of the object version in the context of the Embed
+      //   s3HttpsUri: string -- HTTPS URI of the object version, e.g. https://my-bucket.s3.amazonaws.com/${key}?versionId=${version} (with properly encoded key)
+      //   bucket: string -- current bucket
+      //   key: string -- key of the browsed object
+      //   version: string -- object version id
+      // Example: 'https://my-app/s3-browser?route=<%= encodeURIComponent(url) %>'
+      href?: string
+      // Notification shown after copying href to the clipboard (if `emit` is not set to "override")
+      notification?: string
+      // When set "notify" or "override", Embed will send "s3ObjectLink" message.
```
Comment (akarve) [inline docs/Catalog/Embed.md:229]:
```suggestion
      // When set "notify" or "override", Embed will send the "s3ObjectLink" message.
```
There should be alist of message types, even if it has only one item for now.

File: docs/Catalog/Embed.md [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+  // Bucket name, e.g. 'my-bucket'
+  bucket?: string
+  // Path to object or prefix in the given bucket, e.g. 'some-prefix/some-object.csv'
+  path?: string
+
+  // Initial route Embed will navigate to, e.g. '/b/my-bucket/tree/some/path',
+  // takes precedence over bucket / path
+  route?: string
+
+  // Embed accepts any credentials supported by the Quilt authentication endpoint,
```
Comment (akarve) [inline docs/Catalog/Embed.md:194]: Link to list of support SSO providers https://docs.quiltdata.com/advanced-usage/technical-reference#single-sign-on-sso

File: docs/Catalog/Embed.md [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+
+Catalog also has `/__embed-debug` route which serves a simple debug interface for the Embed:
+
+![](../imgs/embed-debug.png)
+
+It's useful for trying different parameters and inspecting messages passed to/from the Embed.
+
+Its main components are:
+
+1. Inputs for `init` parameters and button for sending `init` command.
```
Comment (akarve) [inline docs/Catalog/Embed.md:343]:
```suggestion
1. Inputs for `init` parameters and button for sending the `init` command.
```

File: docs/Catalog/Embed.md [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+
+2. "Navigate to" button and route input for sending `navigate` command.
```
Comment (akarve) [inline docs/Catalog/Embed.md:345]:
```suggestion
2. "Navigate to" button and route input for sending the `navigate` command.
```

File: docs/Catalog/Embed.md [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+  // Object's key
+  key: string
+  // Object's version ID
+  version: string
+}
+```
+
+## Testing and debugging
+
+Catalog also has `/__embed-debug` route which serves a simple debug interface for the Embed:
```
Comment (akarve) [inline docs/Catalog/Embed.md:335]:
```suggestion
The catalog's `/__embed-debug` route is a simple driver for testing Embed:
```

File: docs/Catalog/Embed.md [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+
+![](../imgs/embed-debug.png)
+
+It's useful for trying different parameters and inspecting messages passed to/from the Embed.
```
Comment (akarve) [inline docs/Catalog/Embed.md:339]:
```suggestion
`__embed-debug` is useful for trying different parameters and inspecting messages passed to and from the Embed.
```

File: docs/Catalog/Embed.md (hunk 6 above)
Comment (nl0) [reply docs/Catalog/Embed.md:229]: other messages are not affected, it's only for `s3ObjectLink`

File: docs/Catalog/Embed.md [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+```
+
+Available commands are listed below.
+
+#### `init`
+
+Initialize the Embed.
+This command must be sent after the Embed is ready
+(see `ready` message reference for details).
+Supported SSO providers are listed in the [Technical Reference](../Technical Reference.md#single-sign-on-sso).
```
Comment (nl0) [inline docs/Catalog/Embed.md:180]: not sure why this doesn't work on gh

Comment (nl0) [reply docs/Catalog/Embed.md:180]: works properly on gitbook tho

**Pull request merged**